src/fluree/db/query/exec/where.cljc

```diff
@@ -22,18 +23,21 @@
 
 (defn resolve-flake-range
   [{:keys [conn t] :as db} error-ch components]
-  (let [out-ch           (async/chan)
-        [s p o]          (map ::val components)
-        [s-fn p-fn o-fn] (map ::fn components)]
+  (let [out-ch               (async/chan)
+        [s-cmp p-cmp o-cmp]  components
+        {s ::val, s-fn ::fn} s-cmp
+        {p ::val, p-fn ::fn} p-cmp
+        {o ::val, o-fn ::fn
+         o-dt ::datatype}    o-cmp]
     (go
       (try* (let [s*          (if (and s (not (number? s)))
                                 (<? (dbproto/-subid db s true))
                                 s)
                   idx         (idx-for s* p o)
                   idx-root    (get db idx)
                   novelty     (get-in db [:novelty idx])
-                  start-flake (flake/create s* p o nil nil nil util/min-integer)
-                  end-flake   (flake/create s* p o nil nil nil util/max-integer)
+                  start-flake (flake/create s* p o o-dt nil nil util/min-integer)
+                  end-flake   (flake/create s* p o o-dt nil nil util/max-integer)
```

Comment:
We support range scans by using [custom comparators](https://github.com/fluree/db/blob/fix/mixed-predicates/src/fluree/db/flake.cljc#L336-L364) which treats non-nil items as equal to `nil`. This allows us to create [flake stubs](https://github.com/fluree/db/blob/fix/mixed-predicates/src/fluree/db/query/exec/where.cljc#L39-L40) containing `nil` for the purposes of matching. For example, if we want to find the value of a predicate for a subject, we do a range scan between flake stubs where the start and end flakes have the subject and values we're interested in in there respective slot, but `nil` as the object, data type, t, and meta. Those `nils` along with our custom comparators allow everything with that subject and predicate to be matched.

The problem arose from us *always* using `nil` for the data type in a query. This works fine when every object value for a given predicate has the same data type (as is usually the case), but when data types were mixed, then specifying an object without it's data type would match both the desired object and *every* other object with a different data type (again, based on how the comparators were implemented to support range scans).

By inferring the data type during the query from the object, we now always specify the data type when we specify the object, so that objects with different data types are automatically rejected.